packages/next/src/client/components/layout-router.tsx

```diff
@@ -114,31 +114,74 @@ function topOfElementInViewport(element: HTMLElement, viewportHeight: number) {
   return rect.top >= 0 && rect.top <= viewportHeight
 }
 
-class ScrollAndFocusHandler extends React.Component<{
+/**
+ * Find the DOM node for a hash fragment.
+ * If `top` the page has to scroll to the top of the page. This mirrors the browser's behavior.
+ * If the hash fragment is an id, the page has to scroll to the element with that id.
+ * If the hash fragment is a name, the page has to scroll to the first element with that name.
+ */
+function getHashFragmentDomNode(hashFragment: string) {
+  // If the hash fragment is `top` the page has to scroll to the top of the page.
+  if (hashFragment === 'top') {
+    return document.body
+  }
+
+  // If the hash fragment is an id, the page has to scroll to the element with that id.
+  return (
+    document.getElementById(hashFragment) ??
+    // If the hash fragment is a name, the page has to scroll to the first element with that name.
+    document.getElementsByName(hashFragment)[0]
+  )
+}
+interface ScrollAndFocusHandlerProps {
   focusAndScrollRef: FocusAndScrollRef
   children: React.ReactNode
-}> {
-  componentDidMount() {
+}
+class ScrollAndFocusHandler extends React.Component<ScrollAndFocusHandlerProps> {
+  handlePotentialScroll = () => {
     // Handle scroll and focus, it's only applied once in the first useEffect that triggers that changed.
     const { focusAndScrollRef } = this.props
 
-    // `findDOMNode` is tricky because it returns just the first child if the component is a fragment.
-    // This already caused a bug where the first child was a <link/> in head.
-    const domNode = findDOMNode(this)
+    if (focusAndScrollRef.apply) {
+      let domNode:
+        | ReturnType<typeof getHashFragmentDomNode>
+        | ReturnType<typeof findDOMNode> = null
+      let isHashScroll = false
+
+      if (focusAndScrollRef.hashFragment) {
+        domNode = getHashFragmentDomNode(focusAndScrollRef.hashFragment)
+        isHashScroll = true
+      }
+
+      // `findDOMNode` is tricky because it returns just the first child if the component is a fragment.
+      // This already caused a bug where the first child was a <link/> in head.
+      if (!domNode) {
+        domNode = findDOMNode(this)
+        console.log('layoutDomNode', domNode)
+      }
+
+      // If there is no DOMNode this layout-router level is skipped. It'll be handled higher-up in the tree.
+      if (!(domNode instanceof HTMLElement)) {
+        return
+      }
 
-    if (focusAndScrollRef.apply && domNode instanceof HTMLElement) {
       // State is mutated to ensure that the focus and scroll is applied only once.
       focusAndScrollRef.apply = false
 
       handleSmoothScroll(
         () => {
+          // In case of hash scroll we need to scroll to the top of the element
+          if (isHashScroll) {
+            window.scrollTo(0, (domNode as HTMLElement).offsetTop)
```

Comment:
a bit annoying that we need this when you have the check above hm

Reply:
Do you mean `as HTMLElement`, yeah somehow it's not picking up the right type, don't want to waste time on it currently.